k8s: Restructure and optimize CEP

tgraf opened the pull request:
This commit creates a native type for the CiliumEndpoint CRD. It still uses a
couple of structures from api/v1/models but the majority of the types are
native. The primary purpose of this commit is to make CEP as stable as possible to avoid frequent updates.

In scope for this PR:
 - Fix structure and sorting of the CEP

Out of scope for this PR:
 - Selective inclusion of certain information based on agent flags or pod annotations.
 - Introduce trigger to update CEP on certain endpoint changes
 - Avoid CEP GET operation for each update interval
 - Avoid k8s version validation for each update interval

## Summary of changes

* CiliumEndpoint.Status.Spec is deprecated and no longer populated. The labe
  configuration is typically available in the pod. The endpoint configuration
  is almost never changed and if performed, it is done via CLI which means the
  CLI can also be used to query it.

* CiliumEndpoint.Status.Status is deprecated but some fields which we know have
  users are preserved. All fields are mirrored to CiliumEndpoint.Status and new
  fields should only be added there.

* The endpoint log is limited to the last 5 errors or warnings. Successful
  transitions are no longer included.

* The list of controllers is limited to failing controllers and the list is
  sorted by controller name.

* The endpoint networking field is reduced and the list of addressess is sorted.

* The policy field is completely restructured and now contains a list of
  identities + L4 port + protocol for ingress and egress.

* As the CEP no longer carries the policy revision, the CI tests are adjusted to solely rely on the CNP status. The CNP status will scale to any number of pods and is what should be the mechanism to determine policy enforcement state. The logic in the Kafka test is kept to validate the policy enforcement state in the CEP.

## Example CEP

```
apiVersion: cilium.io/v2
kind: CiliumEndpoint
metadata:
  creationTimestamp: "2019-01-21T12:20:19Z"
  generation: 1
  name: echo-79b5fcd4d9-2jbxq
  namespace: default
  resourceVersion: "1669965"
  selfLink: /apis/cilium.io/v2/namespaces/default/ciliumendpoints/echo-79b5fcd4d9-2jbxq
  uid: eae1ac79-1d76-11e9-8790-080027d2d952
status:
  external-identifiers:
    container-id: 9241bd4e0b600b46b095376a5120766c726cc7d916e956b6705a89b25cf53dd8
    container-name: k8s_POD_echo-79b5fcd4d9-2jbxq_default_cd4126b4-1d72-11e9-8790-080027d2d952_0
    pod-name: default/echo-79b5fcd4d9-2jbxq
  health:
    bpf: OK
    connected: true
    overallHealth: OK
    policy: OK
  id: 2686
  identity:
    id: 45252
    labels:
    - k8s:io.cilium.k8s.policy.cluster=default
    - k8s:io.cilium.k8s.policy.serviceaccount=default
    - k8s:io.kubernetes.pod.namespace=default
    - k8s:name=echo
  log:
  - code: Warning
    message: 'Skipped invalid state transition to waiting-to-regenerate due to: Triggering
      endpoint regeneration due to '
    state: waiting-to-regenerate
    timestamp: "2019-01-21T12:20:17Z"
  - code: Warning
    message: 'Skipped invalid state transition to ready due to: Completed endpoint
      regeneration with no pending regeneration requests'
    state: waiting-to-regenerate
    timestamp: "2019-01-21T12:20:17Z"
  networking:
    addressing:
    - ipv4: 10.16.195.88
  policy:
    egress:
      allowed:
      - identity: 16777218
        identity-labels:
          cidr:20.1.1.1/32: ""
      - identity: 16777219
        identity-labels:
          cidr:10.128.0.0/9: ""
      - identity: 16777220
        identity-labels:
          cidr:10.0.0.0/10: ""
      - identity: 16777221
        identity-labels:
          cidr:10.64.0.0/11: ""
      - identity: 16777222
        identity-labels:
          cidr:10.112.0.0/12: ""
      enforcing: true
    ingress:
      allowed:
      - identity: 1
        identity-labels:
          reserved:host: ""
      - identity: 2
        identity-labels:
          reserved:world: ""
      - identity: 3
        identity-labels:
          reserved:unmanaged: ""
      - identity: 4
        identity-labels:
          reserved:health: ""
      - identity: 5
        identity-labels:
          reserved:init: ""
      - identity: 100
        identity-labels:
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: cilium-etcd-sa
          k8s:io.cilium/app: etcd-operator
          k8s:io.kubernetes.pod.namespace: ""
      - identity: 101
        identity-labels:
          k8s:app: etcd
          k8s:etcd_cluster: cilium-etcd
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: default
          k8s:io.cilium/app: etcd-operator
          k8s:io.kubernetes.pod.namespace: ""
      - identity: 102
        identity-labels:
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: kube-dns
          k8s:io.kubernetes.pod.namespace: kube-system
          k8s:k8s-app: kube-dns
      - identity: 103
        identity-labels:
          k8s:eks.amazonaws.com/component: kube-dns
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: kube-dns
          k8s:io.kubernetes.pod.namespace: kube-system
          k8s:k8s-app: kube-dns
      - identity: 104
        identity-labels:
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: coredns
          k8s:io.kubernetes.pod.namespace: kube-system
          k8s:k8s-app: kube-dns
      - identity: 105
        identity-labels:
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: cilium-operator
          k8s:io.cilium/app: operator
          k8s:io.kubernetes.pod.namespace: ""
          k8s:name: cilium-operator
      - identity: 106
        identity-labels:
          k8s:eks.amazonaws.com/component: coredns
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: coredns
          k8s:io.kubernetes.pod.namespace: kube-system
          k8s:k8s-app: kube-dns
      - identity: 45252
        identity-labels:
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: default
          k8s:io.kubernetes.pod.namespace: default
          k8s:name: echo
      - identity: 46156
        identity-labels:
          k8s:io.cilium.k8s.policy.cluster: default
          k8s:io.cilium.k8s.policy.serviceaccount: default
          k8s:io.kubernetes.pod.namespace: default
          k8s:name: probe
      - identity: 16777218
        identity-labels:
          cidr:20.1.1.1/32: ""
      - identity: 16777219
        identity-labels:
          cidr:10.128.0.0/9: ""
      - identity: 16777220
        identity-labels:
          cidr:10.0.0.0/10: ""
      - identity: 16777221
        identity-labels:
          cidr:10.64.0.0/11: ""
      - identity: 16777222
        identity-labels:
          cidr:10.112.0.0/12: ""
  spec: {}
  state: ready
  status:
    identity:
      id: 45252
      labels:
      - k8s:io.cilium.k8s.policy.cluster=default
      - k8s:io.cilium.k8s.policy.serviceaccount=default
      - k8s:io.kubernetes.pod.namespace=default
      - k8s:name=echo
    log:
    - code: Warning
      message: 'Skipped invalid state transition to waiting-to-regenerate due to:
        Triggering endpoint regeneration due to '
      state: waiting-to-regenerate
      timestamp: "2019-01-21T12:20:17Z"
    - code: Warning
      message: 'Skipped invalid state transition to ready due to: Completed endpoint
        regeneration with no pending regeneration requests'
      state: waiting-to-regenerate
      timestamp: "2019-01-21T12:20:17Z"
    networking:
      addressing:
      - ipv4: 10.16.195.88
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6697)
<!-- Reviewable:end -->

aanm commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -0,0 +1,228 @@
+			}
+		}
+
+		if controllers != nil {
```
Why are you sorting it on every loop?

tgraf replied:
Oversight, fixed

aanm commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -0,0 +1,228 @@
+	return
+}
+
+func getEndpointStatusLog(e *Endpoint) (log []*models.EndpointStatusChange) {
```
Any reason why it isn't an endpoint method?

tgraf replied:
Good point, changed.

aanm commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -0,0 +1,228 @@
+						log = append(log, l)
+					}
+					added++
+					if added >= cilium_v2.EndpointStatusLogEntries {
```
Please add comment stating what this is doing, it is not clear to me at least

aanm commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -0,0 +1,228 @@
+			Labels: status.Identity.Labels,
+		}
+
+		sort.Strings(identity.Labels)
```
This line is modifying the internal array of `status.Identity.Labels`. Is it guaranteed that `status.Identity.Labels` will not be modified concurrently?

tgraf replied:
Yes but it's not good practice. Added DeepCopy()

aanm commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -0,0 +1,228 @@
+	return
+}
+
+func getEndpointPolicy(e *Endpoint) (policy *cilium_v2.EndpointPolicy) {
```
Could be a method of `Endpoint`

aanm commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -0,0 +1,228 @@
+				}
+			}
+
+			if policy.Ingress.Allowed != nil {
```
Same here, why are you sorting for every loop?

aanm commented on pkg/identity/identity.go:
```diff
@@ -36,6 +36,17 @@ type Identity struct {
+	// reserverd:world
+	//
+	// The CIDRLabel field will only contain 10.0.0.0/8
+	CIDRLabel labels.Labels `json:"-"`
```
If an identity can represent either CIDR or Labels, why can't we reuse one of the existing fields?

tgraf replied:
I don't understand. The Labels field will contains all Labels:
```
        // cidr:10.0.0.0/8
        // cidr:10.0.0.0/7
        // cidr:10.0.0.0/6
        // [...]
        // reserverd:world
```

What field can we reuse?

aanm replied:
ah, I just re-read the comment. In that case, why isn't a CIDRList instead?

tgraf replied:
What should be a CIDRList?

tgraf replied:
The type seems completely unused

aanm replied:
@tgraf the `CIDRLabel`, why isn't the `CIDRLabel` a CIDRList as it only contains CIDRs?

tgraf replied:
```
type CIDRList struct {

        // list
        List []string `json:"list"`

        // revision
        Revision int64 `json:"revision,omitempty"`
}
```

What do we fill as the revision? What is the gain of this additional abstraction?

tgraf replied:
@aanm The CIDRLabel field is exactly *one* label, e.g. `cidr:30.1.1.1/32` or `cidr:10.0.0.0/8`. It is no list. It represents the CIDR identity as the user would expect it.

tgraf replied:
```
      - identity: 16777217
        identityLabels:
          cidr:198.49.23.144/32: ""
```

vs:

```
      - identity: 16777217
        identityLabels:
          cidr:198.49.23.144/32: ""
          cidr:198.49.23.144/31: ""
          cidr:198.49.23.144/30: ""
          cidr:198.49.23.144/29: ""
          cidr:198.49.23.144/28: ""
          cidr:198.49.23.144/27: ""
          cidr:198.49.23.144/26: ""
          cidr:198.49.23.144/25: ""
          cidr:198.49.23.144/24: ""
          cidr:198.49.23.144/23: ""
          cidr:198.49.23.144/22: ""
          cidr:198.49.23.144/21: ""
          cidr:198.49.23.144/20: ""
          cidr:198.49.23.144/19: ""
          cidr:198.49.23.144/18: ""
          cidr:198.49.23.144/17: ""
          cidr:198.49.23.144/16: ""
          cidr:198.49.23.144/15: ""
          cidr:198.49.23.144/14: ""
          cidr:198.49.23.144/13: ""
          cidr:198.49.23.144/12: ""
          cidr:198.49.23.144/11: ""
          cidr:198.49.23.144/10: ""
          cidr:198.49.23.144/9: ""
          cidr:198.49.23.144/8: ""
          cidr:198.49.23.144/7: ""
          cidr:198.49.23.144/6: ""
          cidr:198.49.23.144/5: ""
          cidr:198.49.23.144/4: ""
          cidr:198.49.23.144/3: ""
          cidr:198.49.23.144/2: ""
          cidr:198.49.23.144/1: ""
          "reserved:world": ""
```

aanm replied:
1) how can it be one label? Does it mean we have an identity per CIDR?
2) What about the "except" attribute that we use in the toCIDR? for example:
```
cidr:
  10.0.0.0/8
except:
  10.0.0.127/32
```
how will this translate into an identityLabels?

aanm replied:
also, if it is a single label why not using `labels.Label` instead of `labels.Labels`?

tgraf replied:
> * how can it be one label? Does it mean we have an identity per CIDR?

We have an identity for each CIDR requested by the user. If a user has policy rules which whitelist `10.0.0.0/8` and `10.1.1.1/32` then we have two identities. So it makes sense to report the identity as `cidr:10.0.0.0/8` and `cidr:10.1.1.1/32`

> 2\. What about the "except" attribute that we use in the toCIDR? for example:

Need to double check this but I think this is solved via the identities themselves.

CNP:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "cidr-rule"
spec:
  endpointSelector:
    matchLabels:
      name: echo
  egress:
  - toCIDR:
    - 20.1.1.1/32
  - toCIDRSet:
    - cidr: 10.0.0.0/8
      except:
      - 10.96.0.0/12
```

CEP
```
      - identity: 16777218
        identity-labels:
          cidr:20.1.1.1/32: ""
      - identity: 16777219
        identity-labels:
          cidr:10.128.0.0/9: ""
      - identity: 16777220
        identity-labels:
          cidr:10.0.0.0/10: ""
      - identity: 16777221
        identity-labels:
          cidr:10.64.0.0/11: ""
      - identity: 16777222
        identity-labels:
          cidr:10.112.0.0/12: ""
```

tgraf replied:
> also, if it is a single label why not using `labels.Label` instead of `labels.Labels`?

Mainly to keep the code consistent for both CIDR and non CIDR case. I'm fine either way.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -274,33 +276,253 @@ type CiliumEndpoint struct {
+// and protocol
+type AllowedIdentityTuple struct {
+	Identity       uint64            `json:"identity,omitempty"`
+	IdentityLabels map[string]string `json:"identityLabels,omitempty"`
```
you meant `identity-labels` instead?

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -274,33 +276,253 @@ type CiliumEndpoint struct {
+type AllowedIdentityTuple struct {
+	Identity       uint64            `json:"identity,omitempty"`
+	IdentityLabels map[string]string `json:"identityLabels,omitempty"`
+	DestPort       uint16            `json:"destPort,omitempty"`
```
you meant `dest-port` instead?

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -274,33 +276,253 @@ type CiliumEndpoint struct {
+}
+
+type deprecatedLabelConfigurationSpec struct {
+	User []string `json:"user"`
```
I'm lost, where did this user came from?

tgraf replied:
It's the original structure for backward compatibility:

`api/v1/models/label_configuration_spec.go`
```
type LabelConfigurationSpec struct {

        // Custom labels in addition to orchestration system labels.
        User Labels `json:"user"`
}
```

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -274,33 +276,253 @@ type CiliumEndpoint struct {
+	ID int64 `json:"id,omitempty"`
+
+	// Labels is the list of labels associated with the identity
+	Labels []string `json:"labels"`
```
What's the reason being this being non-omitempty?

tgraf replied:
It's copy paste from the original `api/v1/models/` I think we can omit, yes. Although I don't think it will ever be empty.

tgraf replied:
Added

aanm reviewed (dismissed):
@tgraf it just occurred me, bump the https://github.com/cilium/cilium/blob/fbfc8910bbc845e15c1db22d307194d97aeea288/pkg/k8s/apis/cilium.io/v2/register.go#L44 to 1.13

ianvernon commented on pkg/endpoint/endpoint_status.go:
```diff
@@ -0,0 +1,234 @@
+		networking = &cilium_v2.EndpointNetworking{
+			Addressing: make(cilium_v2.AddressPairList, len(status.Networking.Addressing)),
+		}
+		i := 0
```
i is never incremented , so you will be overwriting prior values in `status.Networking.Addressing`. I can't imagine us ever having more than one IPv4 or IPv6 address, though.

ianvernon commented on pkg/endpoint/endpoint_status_test.go:
```diff
@@ -0,0 +1,251 @@
+	c.ResetTimer()
+	for i := 0; i < c.N; i++ {
+		status := e.GetCiliumEndpointStatus()
+		c.Assert(status, check.Not(check.IsNil))
```
Won't having assertions within a benchmark distort the benchmark a bit?

tgraf replied:
A tiny bit but it's essentially free compared to all of the other operations. The purpose of the benchmark is to give a rough estimate on how many endpoint status operations we can perform per second.

ianvernon reviewed (approved):
Change looks great overall. Are there any concerns with how existing users of CEP in their clusters will be affected? It looks like we are fully backward compatible, but wondering if any other additional testing is needed to ensure that breakage doesn't occur with older CEPs already imported into Kubernetes.

tgraf committed:
k8s: Restructure and optimize CEP

This commit creates a native type for the CiliumEndpoint CRD. It still uses a
couple of structures from api/v1/models but the majority of the types are
native.

* CiliumEndpoint.Status.Spec is deprecated and no longer populated. The labe
  configuration is typically available in the pod. The endpoint configuration
  is almost never changed and if performed, it is done via CLI which means the
  CLI can also be used to query it.

* CiliumEndpoint.Status.Status is deprecated but some fields which we know have
  users are preserved. All fields are mirrored to CiliumEndpoint.Status and new
  fields should only be added there.

* The endpoint log is limited to the last 5 errors or warnings. Successful
  transitions are no longer included.

* The list of controllers is limited to failing controllers and the list is
  sorted by controller name.

* The endpoint networking field is reduced and the list of addressess is sorted.

* The policy field is completely restructured and now contains a list of
  identities + L4 port + protocol for ingress and egress.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test: Adjust CI for new CEP structure

The policy revision is no longer provided in the CEP as updating the CEP that
often does not scale to a large number of pods. Rely on the CNP status instead
which is already verified.

One validation of the status.policy.{ingress|egress}.enforcing is kept to test
the functionality.

Signed-off-by: Thomas Graf <thomas@cilium.io>
